lib/mix/lib/mix/scm/git.ex

```diff
@@ -155,16 +156,25 @@ defmodule Mix.SCM.Git do
     if branch = opts[:branch] do
       "origin/#{branch}"
     else
-      opts[:ref] || opts[:tag] || "origin/master"
+      opts[:ref] || opts[:tag]
     end
   end
 
-  defp get_rev_info do
-    destructure [origin, rev],
-      :os.cmd('git --git-dir=.git config remote.origin.url && git --git-dir=.git rev-parse --verify --quiet HEAD')
+  defp get_rev_info(opts) do
+    cmd = 'git --git-dir=.git config remote.origin.url && git --git-dir=.git rev-parse --verify --quiet HEAD'
```

Comment:
Ok, my concern with the code above is that today it is completely fine to depend on an earlier commit in a branch. For example, imagine you have branch: "master", we don't need to depend on the latest commit in master but this change will always push to the latest commit in there. Therefore, I think the checks we have added are not good, because it is ok to not be in the latest rev in a branch. However, we need to make sure that, whenever you call `mix deps.update`, we will go the latest ref possible in a given branch/tag/rev. I am not sure if we can make it work for `mix deps.get` (which by definition will use whatever is in a lock).